relentless/simulate/lammps.py

```diff
@@ -0,0 +1,870 @@
+    energy_tolerance : float
+        Energy convergence criterion.
+    force_tolerance : float
+        Force convergence criterion.
+    max_iterations : int
+        Maximum number of iterations to run the minimization.
+    dt : float
+        Maximum step size.
+
+    """
+    def __init__(self, energy_tolerance, force_tolerance, max_iterations, dt):
+        self.energy_tolerance = energy_tolerance
+        self.force_tolerance = force_tolerance
+        self.max_iterations = max_iterations
+
+    def to_commands(self, sim):
+        """Calls the appropriate LAMMPS commands for the energy minimization operation.
+
+        Parameters
+        ----------
+        sim : :py:class:`Simulation`
+            The simulation object.
+
+        Returns
+        -------
+        array_like
+            The LAMMPS commands for this operation.
+
+        """
+        cmds = ['minimize {etol} {ftol} {maxiter} {maxeval}'.format(etol=self.energy_tolerance,
+                                                                    ftol=self.force_tolerance,
+                                                                    maxiter=self.max_iterations,
+                                                                    maxeval=100*self.max_iterations)]
+
+        return cmds
+
+class AddLangevinIntegrator(LAMMPSOperation):
+    """Langevin dynamics for a NVE ensemble.
+
+    Parameters
+    ----------
+    dt : float
+        Time step size for each simulation iteration.
+    friction : float or dict
+        Drag coefficient for each particle type (shared or per-type).
+    seed : int
+        Seed used to randomly generate a uniform force.
+
+    """
+    def __init__(self, dt, friction, seed):
+        self.friction = friction
+        self.seed = seed
+
+        self._fix_langevin = self.new_fix_id()
+        self._fix_nve = self.new_fix_id()
+
+    def to_commands(self, sim):
+        """Calls the appropriate LAMMPS commands for the Langevin integrator operation.
+
+        Parameters
+        ----------
+        sim : :py:class:`Simulation`
+            The simulation object.
+
+        Returns
+        -------
+        array_like
+            The LAMMPS commands for this operation.
+
+        Raises
+        ------
+        ValueError
+            If the simulation ensemble does not have constant V.
+        ValueError
+            If the per-type particle masses are not set.
+        KeyError
+            If the friction factor for all types are not set (if per-type
+            friction factors are given).
+
+        """
+        if not sim.ensemble.constant['V']:
+            raise ValueError('Simulation ensemble does not have constant V.')
+
+        # obtain per-type mass (arrays 1-indexed using lammps convention)
+        Ntypes = len(sim.ensemble.types)
+        mass = np.squeeze(sim.lammps.numpy.extract_atom('mass'))
+        if mass is None or mass.shape != (Ntypes+1,):
+            raise ValueError('Per-type masses not set.')
```

Comment:
`np.squeeze` will fail if `extract_atom` returns `None`.
```suggestion
        mass = sim.lammps.numpy.extract_atom('mass')
        if mass is None or mass.shape != (Ntypes+1,1):
            raise ValueError('Per-type masses not set.')
        mass = np.squeeze(mass)
```